docs(client-transcribe-streaming): add readme (#1391)
aws/aws-sdk-js-v3: clients/client-transcribe-streaming/README.md

@aws-sdk/client-transcribe-streaming

[![NPM version](https://img.shields.io/npm/v/@aws-sdk/client-transcribe-streaming/gamma.svg)](https://www.npmjs.com/package/@aws-sdk/client-transcribe-streaming)
[![NPM downloads](https://img.shields.io/npm/dm/@aws-sdk/client-transcribe-streaming.svg)](https://www.npmjs.com/package/@aws-sdk/client-transcribe-streaming)

## Introduction

> Amazon Transcribe streaming transcription enables you to send an audio stream and receive a stream of text in real time.

Reviewer: ```suggestion
Amazon Transcribe streaming enables you to send an audio stream and receive back a stream of text in real time.
```

installation. After installation, in the `index.js`, you can import the Transcribe Streaming client like:

```javascript
const { TranscribeStreamingClient, StartStreamTranscriptionCommand } = require("@aws-sdk/client-transcribe-streaming");
```

If `require` is not available on the platform you are working on(browsers). You can import the client like:

```javascript
import { TranscribeStreamingClient, StartStreamTranscriptionCommand } from "@aws-sdk/client-transcribe-streaming";
```

## Constructing the Service Client

You can create a service client like bellow:

```javascript
const client = new TranscribeStreamingClient({
  region,
  credentials,
});
// region and credentials are optional in Node.js
```

## Acquire Speech Stream

The Transcribe Streaming client accepts streaming speech input as an [async iterable]. You can construct them from either an async generator or using `Symbol.asyncIterable` to emit binary chunks.

Here's an example of using async generator:

```javascript
const audioStream = async function* () {
  await device.start();
  while (device.ends !== true) {
    const chunk = await device.read();
    yield chunk; /* yield binary chunk */
  }
};
```

> Then you need to construct the binary chunk into the audio chunk shape that can be recognized by the SDK:

Reviewer: ```suggestion
Then you need to construct the binary chunk into an audio chunk shape that can be recognized by the SDK:
```

```javascript
const audioStream = async function* () {
  for await (const chunk of audioSource()) {
    yield { AudioEvent: { AudioChunk: chunk } };
  }
};
```

### Acquire from Node.js Stream API

In Node.js you will mostly acquire the speech in [Stream API], from HTTP request or devices. Stream API in Node.js (>=

> 10.0.0) itself is an async iterable. You can supply the the streaming into the SDK input without explicit convert. You

Reviewer 2: ```suggestion
10.0.0) itself is an async iterable. You can supply the streaming into the SDK input without explicit convert. You
```

  // high-quality audio. The sample rate must match the sample rate in the audio file.
  MediaSampleRateHertz: 44100,
  AudioStream: audioStream(),
});
const response = await client.send(command);
```

## Handling Text Stream

If the request succeeds, you will get a response containing the streaming transcript like [this][response data example].
Just like the input speech stream, the transcript stream is an [async iterable][async iterable] emitting the partial
transcripts. Here is a code snippet of accessing the transcripts

```javascript
// This snippet should be put into an async function
for await (const event of response.TranscriptResultStream) {
  if (event.TranscriptEvent) {
    const message = event.TranscriptEvent;
    // Get multiple possible results
    const results = event.TranscriptEvent.Transcript.Results;
    // Print all the possible transcripts
    results.map((result) => {
      (result.Alternatives || []).map((alternative) => {
        const transcript = alternative.Items.map((item) => item.Content).join(" ");
        console.log(transcript);
      });
    });
  }
}
```

### Pipe Transcripts Stream

> In Node.js, you can pipe this `TranscriptResultStream` other destination easily with the [`from` API][readable.from]:

Reviewer: ```suggestion
In Node.js, you can pipe this `TranscriptResultStream` to other destinations easily with the [`from` API][readable.from]:
```

```javascript
const { Readable } = require("stream");
const transcriptsStream = Readable.from(response.TranscriptResultStream);
transcriptsStream.pipe(/* some destinations */);
```

## Error Handling

If you are using `async...await` style code, you are able to catch the errors with `try...catch` block. There are 2

> categories of exceptions can be thrown: Immediate exceptions thrown before transcription is started, like signature
> exceptions, invalid parameters exceptions, and network errors; Streaming exceptions that happens after transcription is
> started, like `InternalFailureException` or `ConflictException`.

Reviewer 2: Use list
```suggestion
categories of exceptions can be thrown:

- Immediate exceptions thrown before transcription is started, like signature
exceptions, invalid parameters exceptions, and network errors;
- Streaming exceptions that happens after transcription is
started, like `InternalFailureException` or `ConflictException`.
```

> For the first category, the SDK client will retry the request if the error is retryable, like network errors. You can

Reviewer 2: ```suggestion
For immediate exceptions, the SDK client will retry the request if the error is retryable, like network errors. You can
```

> config the client to behave as you intend to. For the second category, because the streaming transcription is already

Reviewer 2: ```suggestion
config the client to behave as you intend to.

For Streaming exceptions, because the streaming transcription is already
```

started, client cannot retry the request automatically. The client will throw these exceptions and users can handle the
stream behavior accordingly.

Here's an example of error handling flow:

```javascript
const { InternalFailureException, ConflictException } = require("@aws-sdk/client-transcribe-streaming");
try {
  const response = await client.send(command);
  await handleResponse(response);
} catch (e) {
  if (InternalFailureException.isa(e)) {
    /* handle InternalFailureException */
  } else if (ConflictException.isa(e)) {
    /* handle ConflictException */
  }
} finally {
  /* clean resources like input stream */
}
```

## Notes for React Native

This package is compatible with React Native (>= 0.60). However, it is not tested with any React Native libraries that
converts microphone record into streaming data. Community input for integrating streaming microphone record data is

> welcome

Reviewer 2: ```suggestion
welcome.
```

Thank you for reading this guide. If you want to know more about how streams are encoded, how connection is established,

> please refer to the [Service API guide][service api guide]

Reviewer 2: ```suggestion
please refer to the [Service API guide][service api guide].
```
